src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -94,27 +93,74 @@ internal Binder TypeSyntaxBinder
+        /// which can be inferred (if necessary) by binding the enclosing statement.
+        /// </summary>
         public static SourceLocalSymbol MakeDeconstructionLocal(
             Symbol containingSymbol,
-            Binder binder,
+            Binder scopeBinder,
             TypeSyntax closestTypeSyntax,
             SyntaxToken identifierToken,
             LocalDeclarationKind kind,
             SyntaxNode deconstruction)
         {
             Debug.Assert(closestTypeSyntax != null);
 
+            Debug.Assert(closestTypeSyntax.Kind() != SyntaxKind.RefType);
             if (closestTypeSyntax.IsVar)
             {
                 return new DeconstructionLocalSymbol(
-                    containingSymbol, binder, closestTypeSyntax, identifierToken, kind, deconstruction);
+                    containingSymbol, scopeBinder, closestTypeSyntax, identifierToken, kind, deconstruction);
             }
             else
             {
-                return new SourceLocalSymbol(containingSymbol, binder, false, closestTypeSyntax, identifierToken, kind);
+                return new SourceLocalSymbol(containingSymbol, scopeBinder, false, closestTypeSyntax, identifierToken, kind);
             }
         }
 
+        /// <summary>
+        /// Make a pattern local variable symbol which can be inferred (if necessary) by binding the enclosing pattern-matching operation.
+        /// </summary>
+        internal static LocalSymbol MakePatternLocalSymbol(
+            Symbol containingSymbol,
+            Binder scopeBinder,
+            DeclarationPatternSyntax node)
+        {
+            Debug.Assert(node.Type.Kind() != SyntaxKind.RefType);
+            if (!node.Type.IsVar)
+            {
+                return new SourceLocalSymbol(containingSymbol, scopeBinder, false, node.Type, node.Identifier, LocalDeclarationKind.PatternVariable);
+            }
+
+            //
+            // A pattern variable's type can be inferred from context.
+            //
+            // The syntax allows a pattern either (1) on the right-hand-side of an "is-pattern" expression, or
+            // (2) in the pattern of a switch case. These have their type inferred using different mechanisms,
+            // so we need to find that context. These are the only cases the parser will produce, so we can be
+            // confident that we will find one of them by scanning up node's .Parent chain. We allow
+            // a NullReferenceException at n.Parent to occur when we are fed bad trees.
+            //
+            for (SyntaxNode n = node; ; n = n.Parent)
+            {
+                var kind = n.Kind();
+                if (kind == SyntaxKind.IsPatternExpression)
+                {
+                    var expr = (IsPatternExpressionSyntax)n;
+                    return new VariableDeclaredInExpression(containingSymbol, scopeBinder, scopeBinder, node.Type, node.Identifier, LocalDeclarationKind.PatternVariable, expr);
```

Comment:
Yes, as we do not have any binders that can change between the `e` and `x` in the expression `e is var x`.

---
In reply to: [75760180](https://github.com/dotnet/roslyn/pull/13274#discussion_r75760180) [](ancestors = 75760180)